psc-flink/src/main/java/com/pinterest/flink/streaming/connectors/psc/table/PscRateLimitMap.java

```diff
@@ -0,0 +1,120 @@
+ * to a specified rate. The total rate limit is divided evenly across all
+ * parallel subtasks, ensuring the aggregate rate does not exceed the configured limit.
+ *
+ * <p>When the rate limit is exceeded, the function blocks until permits become available,
+ * emitting metrics to track throttling behavior.
+ *
+ *
+ *
+ * @param <T> The type of records flowing through this map function
+ */
+public class PscRateLimitMap<T> extends RichMapFunction<T, T> {
+
+    private static final long serialVersionUID = 1L;
+
+    private final double rateLimitRecordsPerSecond;
+    private transient Counter throttleOccurredCounter;
+    private transient Gauge<Long> maxThrottleDelayGauge;
+    private transient RateLimiter subtaskRateLimiter;
+    private volatile long maxThrottleDelayMs = 0;
+
+    /**
+     * Creates a new rate limiting map function.
+     *
+     * @param rateLimitRecordsPerSecond 
+     */
+    public PscRateLimitMap(double rateLimitRecordsPerSecond) {
+        this.rateLimitRecordsPerSecond = rateLimitRecordsPerSecond;
+    }
+
+    @Override
+    public void open(Configuration configuration) throws Exception {
+        super.open(configuration);
+
+        // Register metrics
+        throttleOccurredCounter = getRuntimeContext()
+                .getMetricGroup()
+                .counter("throttleOccurred");
+
+        maxThrottleDelayGauge = getRuntimeContext()
+                .getMetricGroup()
+                .gauge("maxThrottleDelayMs", () -> maxThrottleDelayMs);
+
+        // Calculate per-subtask rate limit
+        int numberOfParallelSubtasks = getRuntimeContext().getNumberOfParallelSubtasks();
+        double subtaskRateLimit = rateLimitRecordsPerSecond / numberOfParallelSubtasks;
+
+        Preconditions.checkArgument(
+                subtaskRateLimit > 0.1,
+                "Subtask rate limit should be greater than 0.1 QPS. " +
+                        "Current rate: %s records/second divided by %s subtasks = %s records/second per subtask. " +
+                        "Consider increasing the rate limit or decreasing parallelism.",
+                rateLimitRecordsPerSecond,
+                numberOfParallelSubtasks,
+                subtaskRateLimit);
+
+        subtaskRateLimiter = RateLimiter.create(subtaskRateLimit);
+    }
+
+    @Override
+    public T map(T record) throws Exception {
+        blockIfNecessary();
+        return record;
+    }
+
+    /**
+     * Blocks if necessary to enforce the rate limit.
+     * Emits metrics when throttling occurs.
+     */
+    private void blockIfNecessary() {
+        // tryAcquire returns false if there are no permits available at this time
+        if (!subtaskRateLimiter.tryAcquire()) {
+            // Record that throttling occurred
+            throttleOccurredCounter.inc();
+            long startTime = System.currentTimeMillis();
+
+            // Block until a permit becomes available
+            subtaskRateLimiter.acquire();
```

Comment:
Side note that we can discuss outside this PR: This might be the key reason why most `RateLimiter` operators in our pipelines are read on CPU during busy times. We should probably think about an async way to handle it.